subsys/modbus/modbus_serial.c

```diff
@@ -393,7 +393,7 @@ static void cb_handler_tx(struct modbus_context *ctx)
 				   cfg->uart_buf_ctr);
 		cfg->uart_buf_ctr -= n;
 		cfg->uart_buf_ptr += n;
-	} else {
+	} else if (uart_irq_tx_complete(cfg->dev) != 0) {
```

Comment:
Hello,
I have opened an issue for this under #40578 rather than a new PR as there might be other boards affected by this. Developers submitting PR's referencing the issue and fixing their respective platforms rather than messaging me to add their changes seems like a better option.

Regarding the fix for modbus, it was tested at the physical layer with the modbus driver only and not with any of the zephyr tests. It should be noted that the tests themselves could be broken such as was the case on PR's I previously submitted #38085 which was fixed in #35721

This fix works for sam0 but since it appears other platforms require verification, other developers should test their respective platforms and verify on physical hardware with logic analyzers or oscilloscopes.